build.gradle.kts

```diff
@@ -96,55 +95,60 @@ fun shouldPublishSnapshots(): Boolean {
   return eventName == "push" && (ref == "refs/heads/main" || ref == "refs/heads/dev-3.x")
 }
 
-tasks.register("publishSnapshotsIfNeeded") {
+tasks.register("ciPublishSnapshot") {
+  description = "Publishes a SNAPSHOT"
+
   if (shouldPublishSnapshots()) {
-    project.logger.log(LogLevel.LIFECYCLE, "Deploying snapshot to OSS Snapshots...")
     dependsOn(subprojectTasks("publishAllPublicationsToOssSnapshotsRepository"))
+  } else {
+    doFirst {
+      error("We are not on a branch, fail snapshots publishing")
+    }
   }
 }
 
-tasks.register("publishToOssStagingIfNeeded") {
+
+tasks.register("ciPublishRelease") {
+  description = "Publishes all artifacts to OSSRH and the Gradle Plugin Portal"
+
   if (isTag()) {
-    project.logger.log(LogLevel.LIFECYCLE, "Deploying release to OSS staging...")
     dependsOn(subprojectTasks("publishAllPublicationsToOssStagingRepository"))
+    // Only publish plugins to the Gradle portal if everything else succeeded
+    finalizedBy(":apollo-gradle-plugin:publishPlugins")
+  } else {
+    doFirst {
+      error("We are not on a tag, fail release publishing")
+    }
   }
+
 }
 
-tasks.register("publishToGradlePortalIfNeeded") {
-  if (isTag()) {
-    project.logger.log(LogLevel.LIFECYCLE, "Deploying release to Gradle Portal...")
-    dependsOn(":apollo-gradle-plugin:publishPlugins")
-  }
+tasks.register("ciTestsGradle") {
+  description = "Execute the Gradle tests (slow)"
+  dependsOn(":apollo-gradle-plugin:test")
 }
 
-tasks.register("fullCheck") {
+tasks.register("ciTestsNoGradle") {
+  description = """Execute all tests from the root project except: 
+    | - the Gradle ones
+    | - most of the apple tests where it executes only macosX64
+  """.trimMargin()
+
+
   subprojects {
-    tasks.all {
-      if (this.name == "build") {
-        this@register.dependsOn(this)
-      }
+    if (name != "apollo-gradle-plugin") {
+      dependsOn(tasks.matching { it.name == "test" })
     }
+    dependsOn(tasks.matching { it.name == "jvmTest" })
```

Comment:
That's using an eager API that defeats configuration avoidance FYI.

Reply:
Oh no, not again 😅 . I thought `TaskCollection.matching {}` was ok? What is the eager part?

Reply:
That's what I recall reading from @melix a few days ago. Maybe there's been come confusion?

Reply:
`matching` is eager. `tasks.withType(...)` is not. So it's better to use `tasks.named("jvmTest", Test)` which is lazy.

Reply:
Wait, I'm confused now, the page I linked above explicitly says about `TaskCollection.matching(groovy.lang.Closure)`:

```
This is OK to use because it does not require tasks to be created immediately.
```

What am I missing?

Reply:
`tasks.matching { ... }` is not immediately realizing the collection. However, as soon as it is queried, it requires the realization of the tasks it filters. So the `dependsOn(...)` will trigger the realization of the tasks.

If you had done `tasks.withType(Test).matching { ... }`, then the only realized tasks would be of type `Test`.

Reply:
That might take us a bit too far but is there a reason matching on `task.name` requires realization while matching on `task.javaClass` doesn't? Looks like there's no way to depend on a task that might or might not be already registered while only knowing its name? 

In the context of this PR, this is most likely not a huge issue as the `dependsOn` is made when configuring a task that's only meant to be called from CI. If everything goes well, chances are that only CI is affected 🤞 . In the long run, we might benefit from something like https://developer.squareup.com/blog/herding-elephants/ so that we can more precisely add the dependencies to our CI tasks (and not have to iterate all subprojects and all tasks as this PR does). Follow up issue there: https://github.com/apollographql/apollo-android/issues/3531

Reply:
`tasks.matching` is an old API, largely predates lazy configuration. Therefore it doesn't match on the _prototypes_ of tasks (name, type) but on the _realized_ tasks. The `withType` is more recent and creates a view which only requires the types. Similarly, `named` only requires the names, so `named(name, type)` works lazily in both cases (because it only needs the prototype which is used in `register`).